Add emulators for extended cosmological models

suicee opened the pull request.

DeaglanBartlett commented:
Thank you for adding these @suicee! Please could you add an example to the examples folder to demonstrate how to use these functions and to compare against, e.g., the results of class (or class+halofit for the nonlinear case)?

suicee commented:
I've made the following updates:
1. Added the PyTorch implementation.
2. Included example usage for the emulators and their PyTorch version in `examples/siren_plus_example.ipynb`.
3. Added `__init__.py` files to facilitate module imports.

Let me know if you have any suggestions, and feel free to modify anything that you think could be improved.

DeaglanBartlett reviewed (changes requested):
This is all really good! I have left a few comments, which mainly relate to code formatting so should be easy to fix. 

Given we now have a torch implementation of some files (the extended cosmology), I wonder whether we can make the same for the standard cosmology. I think the best way of dealing with this is if you put the `logF` function in a new file `linear.py` and I will make another issue to produce the remaining code.

suicee commented:
I've updated the docstrings for most functions; please let me know if there’s anything I missed. However, I haven’t implemented your suggestion to import  `logF_fiducial` from `linear.py`, because you've already defined handling for out-of-range $k$ values within that module. We may want to consider handling out-of-range cases independently within `linear_plus.py` if needed. Additionally, in your current implementation, $k=9$ is out of range, which may cause some of my examples to break.

DeaglanBartlett commented:
I have made the required changes such that `linear_plus` calls `logF_fiducial` from `linear` but the out-of-range checks are not called so that the behaviour is unchanged compared to @suicee's code but without the duplicate code.

DeaglanBartlett commented on symbolic_pofk/pytorch/linear_plus.py:
```diff
@@ -0,0 +1,288 @@
+
+    Pk = eh * D ** 2 * F_value * R_value * S_value
+
+    return Pk.T
```
When running this code, I get the following warning:

```
UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/native/TensorShape.cpp:3641.)
```

What is the behaviour you want? Please can you choose the appropriate one and change to it?
